---
title: "R Notebook"
output: html_notebook
---
Before starting make sure to set the working directory to this folder "BIMM143----Project-3-main" before running code, or else R can not find the FASTA files.
_____________________________________________________________________________________
Scientific Question: Will 10(Z)-hexadecenoic acid produce a similar anti-inflammatory response in human macrophages compared to mouse macrophages?

Isolated from Mycobacterium vaccae, 10(Z)-hexadecenoic acid was determined to be the compound responsible for the anti-inflammatory and immune suppression effects that were observed when injecting heat-killed M. vaccae into mice (Zuany et al. 2002, Lowry et al. 2017). The anti-inflammatory effects of 10(Z)-hexadecenoic acid in mice specifically comes from its role as an agonist to peroxisome proliferator-activated receptor alpha (PPARA), which results in less proinflammatory cytokines, such as Interleukin-6, being produced following stimulation by lipopolysaccharide (LPS), a molecule commonly found on bacteria (Smith et al. 2019). The PPARA nucleotide sequences (coding sequence only) and amino acid sequences were obtained from NCBI and UniProt respectively and chosen after searching on the UCSC genome browser, with their IDs listed as the name of the FASTA file. 

_____________________________________________________________________________________
Scientific Hypothesis: If the PPARA of mice and humans are highly similar, then the anti-inflammatory effects of 10(Z)-hexadecenoic acid observed in mice should also be observed in humans. 

Here we perform pairwise alignments of the nucleotide sequences and of the amino acid sequences of mouse and human PPARA, and visualize the alignment using a dotplot. Pairiwise alignment scores of the two pairs of sequences will be used to estimate sequence similarity quantitatively, while the dotplot of the two pairs of sequences will be used to estimate sequence similarity qualitatively. 

_____________________________________________________________________________________
The following are necessary packages that need to be installed and loaded into R: Biostrings, seqinr.
Biostrings is needed to make pairwise alignments, while seqinr is needed to load the data in and make the dotplots.
To do so, simply type the following commands:
```{r}
#Load the packages Biostrings and seqinr, which will be used for analyzing and visualizing the DNA and protein sequence similarity
require(Biostrings)
require(seqinr)
```

_____________________________________________________________________________________
Below is the code to load in the FASTA files of the nucleotide sequences (only coding sequence of mRNA) and the amino acid sequences. One file is printed out for visualization of the data. Make sure to set the working directory to the folder "BIMM143----Project-3-main" before running file, or else R can not find the FASTA files.
```{r}
#Load the mRNA and amino acid sequences to R using read.fasta, a function in seqinr package.
#Note that the mRNA files use DNA nucleotides, and thus are cDNA. I specifically used coding sequences (CDS) as it contains only the exons, which yields the same length "mRNA". I will continue referring to the file as "mRNA" to be consistent with the source where I got the FASTA files from.
mPPARARNA <- read.fasta(file = "NM_001113418.1.fasta") #Mouse PPARA "mRNA" (transcript variant 2 from NCBI Nucleotide, 1407 nt)
hPPARARNA <- read.fasta(file = "NM_001001928.4.fasta") #Human PPARA "mRNA" (transcript variant 2 from NCBI Nucleotide, 1407 nt)
mPPARAPro <- read.fasta(file = "P23204.fasta", seqtype = "AA") #Mouse PPARA Protein (from UniProt, 468 AA)
hPPARAPro <- read.fasta(file = "Q07869.fasta", seqtype = "AA") #Human PPARA Protein (isoform 1 from UniProt, 468 AA)

#Data print out of mouse PPARA "mRNA"
mPPARARNA
class(mPPARARNA)
```

Notice that the "mRNA" sequence uses DNA nucleotides, is in lower case letters, and is in a list.
Before we can move on, we actually need to manipulate the files so that the functions can read the data.
The Biostrings functions pairwise alignment reads single string Biostrings objects, while seqinr function reads vectors of single characters. We can accomplish these respective conditions using DNAString and list subsetting.
```{r}
#Making Biostrings for pairwise alignments.
mouseDNAseq <- DNAString(c2s(mPPARARNA[[1]])) #c2s converts the vector of single characters to a single string, which is read by DNAString function and converted to a Biostring. 
humanDNAseq <- DNAString(c2s(hPPARARNA[[1]]))
mouseAAseq <- AAString(c2s(mPPARAPro[[1]]))
humanAAseq <- AAString(c2s(hPPARAPro[[1]]))
#We have successfully made a Biostrings object!
class(mouseDNAseq)
mouseDNAseq
```
```{r}
#Making vector of single characters for dotplot
mPPARARNAseq <- mPPARARNA[[1]] #Mouse PPARA "RNA" sequence in character vector format
hPPARARNAseq <- hPPARARNA[[1]] 
mPPARAProseq <- mPPARAPro[[1]] 
hPPARAProseq <- hPPARAPro[[1]]
#We have successfully made a character vector of class SeqFastadna (or SeqFastaAA for amino acid sequence) 
class(mPPARARNAseq)
mPPARARNAseq
```

In the next chunk of code, we will be doing the pairwise alignments. 
First we must generate the scoring matrices for the matchings. For the amino acid sequences, there are premade scoring matrices made called BLOSUM. The higher BLOSUM numbers are used for more related sequences, such as BLOSUM80, while lower numbers are used for more distant sequences, such as BLOSUM45.
```{r}
#Load the scoring matrices used for the nucleotide pairwise alignments from the Biostrings package. 
DNAScoreMatrix <- nucleotideSubstitutionMatrix(match = 2, mismatch = -1, baseOnly = TRUE) #Load DNA scoring matrix
DNAScoreMatrix
data(BLOSUM80) #Load AA scoring matrix
BLOSUM80

#Conduct pairwise alignment of the human and mouse PPARA nucleotide and amino acid sequences.
pairAlignsmhDNAseq <- pairwiseAlignment(mouseDNAseq, humanDNAseq, substitutionMatrix = DNAScoreMatrix, gapOpening = -2, gapExtension = -8, scoreOnly = FALSE)
pairAlignsmhDNAseq #Alignment score of mouse and human PPARA DNA sequences.
pairAlignsmhAAseq <- pairwiseAlignment(mouseAAseq, humanAAseq, substitutionMatrix = "BLOSUM80", gapOpening = -2, gapExtension = -8, scoreOnly = FALSE)
pairAlignsmhAAseq #Alignment score of mouse and human PPARA AA sequences.
```

We can see that the pairwise alignment score for the two DNA sequences is 2205 and the score for the two amino acid sequences is 2485. 
However, the alignment score alone is not very useful without anything to compare to. We can calculate the max score by multiplying the matches by the match score, and then compare the two values to draw conclusions about the quantitative similarity of the sequences. 
For the DNA sequences, the max score (full match) is simply the number of nucleotides (1407) times the score for each match (2).
For the amino acid sequences, the max score (full match) is the number of each amino acid present times the score for each match (values can be found in BLOSUM80 matrix). Since the amino acids each have different match scores, I will take the average between the human and mouse max scores, which are similar in value, to compare to the alignment score
```{r}
#Calculating nucleotide sequence max score.
DNAMaxScore <- length(mouseDNAseq)*2 #The mouse and human DNA sequences are the same length, so either will work. We multiply by 2 because each match is a score of 2 in the pairwise alignment.
cat("If the DNA sequences fully match each other, the max score will be ", DNAMaxScore, "\n") #Print out of DNA max score.

#Calculating amino acid sequence max score.
AAMatchScores <- c(5, 6, 6, 6, 9, 6, 6, 6, 8, 5, 4, 5, 6, 6, 8, 5, 5, 11, 7, 4, 0, 0, 5, 3, 5, -1, 1, 0, 0, 0, 0) #BLOSUM80 match scores
mAAfreq <- alphabetFrequency(mouseAAseq) #AA frequency count of mouse sequence.
hAAfreq <- alphabetFrequency(humanAAseq) #AA frequency count of human sequence.
mouseAAMaxScore <- sum(mAAfreq*AAMatchScores) #Calculation of the max score (full match) to mouse AA sequence.
cat("If the amino acid sequences fully match to the mouse sequence, the max score will be ", mouseAAMaxScore, "\n") #Score if fully matches mouse AA sequence.
humanAAMaxScore <- sum(hAAfreq*AAMatchScores) #Calculation of the max score (full match) to human AA sequence.
cat("If the amino acid sequences fully matches to the human sequence, the max score will be ", humanAAMaxScore, "\n") #Score if fully matches human AA sequence.
mhAvgAAMaxScore <- (mouseAAMaxScore + humanAAMaxScore)/2 #Average of the two scores
cat("The average max amino acid score will be ", mhAvgAAMaxScore, "\n") #Print out of average AA max score.
```

Now we can compare to the pairwise alignment score, and estimate a percent similarity.
```{r}
#Reminder of what the alignment scores are.
pairAlignsmhDNAseq #DNA alignment score, which is 2205.
pairAlignsmhAAseq #AA alignment score, which is 2485.

#Calculate percent similarity.
mhDNAsim <- 2205/DNAMaxScore*100
cat("The mouse and human DNA sequences are about ", round(mhDNAsim),"% similar", "\n")
mhAAsim <- 2485/mhAvgAAMaxScore*100
cat("The mouse and human amino acid sequences are about ", round(mhAAsim),"% similar", "\n")
```

We can see that the DNA percent similarity is about 78%, while the AA percent similarity is about 94%. 
Next, we can visualize the the alignment using the dotplot function of the seqinr package.
```{r}
#Visualize sequence alignment using dotplot 
dotPlot(mPPARARNAseq, hPPARARNAseq, wsize = 9, wstep = 9, nmatch = 8, xlab = 'Mouse DNA sequence', ylab = 'Human DNA sequence') #Nucleotide similarity visualization
dotPlot(mPPARAProseq, hPPARAProseq, wsize = 5, wstep = 5, nmatch = 4, xlab = 'Mouse Amino Acid Sequence', ylab = 'Human Amino Acid Sequence') #Amino acid similarity visualization
```
Using a big window and high stringency, we can see a very distinct line in both the DNA and amino acid sequence dotplots, indicating that the mouse and human PPARA DNA sequences are similar and the mouse and human amino acid sequences are also similar. There are much more gaps in the line of the DNA dotplot compared to the amino acid dotplot, which is expected because the DNA sequences are not as similar (~78%) compared to the amino acid sequences (~94%).

_____________________________________________________________________________________

Analysis: While the PPARA DNA sequence percent similarity of mice and humans is somewhat low (78%), the PPARA amino acid sequence percent similarity is high (94%), and these results are somewhat reflective in the dotplots, where the DNA dotplot has more gaps compared to the amino acid dotplot. Although the DNA similarity is relatively low, I would still conclude that the human and mouse PPARA receptors are very similar because different combinations of 3 RNA bases can code for the same amino acid, and thus yield the same overall amino acid despite the DNA or RNA sequences being different. In addition, protein function and chemistry is determined by the amino acid sequences, and thus since the sequences are highly similar, it is very likely that the human and mouse PPARA have similar chemistry. Thus, 10(Z)-hexadecenoic acid should have a similar agonistic effect on human PPARA receptors like it does on mouse PPARA receptors, and produce an anti-inflammatory response in human macrophages that is similar to mouse macrophages.



